### PR TITLE
Problem: no persistence of WalletConnect session state (fix #3)

### LINKED ISSUE #3: Problem: no persistence of WalletConnect session state

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: add callback for session state
- set_callback  to receive directly
- run_callback to receive through callback function

### CODE REVIEW

File: wallet-connect/src/client.rs
Comment (tomtau) [inline wallet-connect/src/client.rs]: what about if the client state gets simply updated -- e.g. the user changes the default address or network in the wallet?

**Review by tomtau — COMMENTED** [review]
how will the caller be able to restore the session information after shutting down process and starting it up later: https://github.com/crypto-com/game-sdk-extras/blob/main/wallet-connect/src/client/session.rs#L18 ?

For example, how will it know what addresses the wallet provided? The state message sent in callbacks seems to only contain Uri which has a part of the information (secret key, bridge server URL, handshake topic...), but not all -- i.e. the caller can't save/persist the session state?

File: wallet-connect/src/client.rs
Comment (leejw51crypto) [reply wallet-connect/src/client.rs]: ClientState is for delivering data through channel(read only), so user notified with ClientState.

working on these
1. save & load session(restroing from serialization) , user can store anywhere
2. in apply update in session, deliver session info to the user via callback, user refreshes, and maybe save again.
3. after restoring session, and ensure_session will work as it is, (in next try, shutdown current one, start the program again)
so it's persistent, user just use revived session to await for signature.

Comment (leejw51crypto) [reply wallet-connect/src/client.rs]: added save , load session info, restoration added

but after restoration, signing not working. solving now.

File: wallet-connect/examples/web3.rs
Comment (elecming) [inline wallet-connect/examples/web3.rs]: It doesn't seem to work. When I disconnect, it throws error:

```
Error: missing field `approved`

Location:
    /Users/damonchen/Documents/game-sdk-extras/wallet-connect/src/client/socket.rs:181:30

```

Comment (elecming) [reply wallet-connect/examples/web3.rs]: And how to test the persistent session feature?

Comment (leejw51crypto) [reply wallet-connect/examples/web3.rs]: ok , i'll fix

**Comment by leejw51crypto** [comment]
now, persistent, saved to session.info,
after successful signing, session.info is removed

and fix `approve` parsing error

File: wallet-connect/src/client.rs
Comment (tomtau) [inline wallet-connect/src/client.rs]: `uri` seems redundant? It can be reconstructed from the session info -- in the `Session`, there's a method `fn uri` which can construct it?

File: wallet-connect/src/client.rs
Comment (tomtau) [inline wallet-connect/src/client.rs]: maybe a different name or some comment for explanation what this sender is for, because there are other internal senders in the connector/socket etc.

**Comment by leejw51crypto** [comment]
fixing ci errors

File: wallet-connect/src/client/session.rs
Comment (tomtau) [inline wallet-connect/src/client/session.rs]: it may not be so -- see line 98... the update may be to be disconnected

File: wallet-connect/src/client.rs
Comment (tomtau) [inline wallet-connect/src/client.rs]: is this `ClientState` also redundant? `connected` is contained inside `SessionInfo`?

File: wallet-connect/src/client.rs
Comment (tomtau) [inline wallet-connect/src/client.rs]: is it ok to unwrap?

File: wallet-connect/src/client/core.rs
Comment (tomtau) [inline wallet-connect/src/client/core.rs]: I'm not sure why this is needed? `Uri` remains the same from the previously restored session -> the handshake topic should be the same.

just for reference, the original reference code (it seems it just switched to ethers as well) has persistence and doesn't invalidate handshake topics: https://github.com/nlordell/walletconnect-rs and there isn't anything about it in the technical spec

File: wallet-connect/src/client/session.rs
Comment (tomtau) [inline wallet-connect/src/client/session.rs]: safe to unwrap?

**Review by tomtau — CHANGES_REQUESTED** [review]
I think the persistence should work now, but two potential bugs:
1. removing handshake topics on restore
2. not notifying the client of wallet disconnection update due to duplicate data

There are also quite a few `clone` calls that may be possible to remove, but that's just a nit

File: wallet-connect/src/client.rs
Comment (leejw51crypto) [reply wallet-connect/src/client.rs]: ok

File: wallet-connect/src/client.rs
Comment (leejw51crypto) [reply wallet-connect/src/client.rs]: ok, i'll change the name

File: wallet-connect/src/client/session.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/session.rs]: ok

**Comment by leejw51crypto** [comment]
> I think the persistence should work now, but two potential bugs:
> 
> 1. removing handshake topics on restore
> 2. not notifying the client of wallet disconnection update due to duplicate data
> 
> There are also quite a few `clone` calls that may be possible to remove, but that's just a nit

tried to use same topic, but user approval never delivered, but next try, it's delievered.
so changed to invalidate old handshake topic.

File: wallet-connect/src/client/session.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/session.rs]: it's unbounded channel, never fails, so i'll change expect ,and comment

File: wallet-connect/src/client/core.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/core.rs]: yes, i'll check again. 
somehow persistent handshake not working correctly, user approval not delivered.
possibly, once subscribed, and disconencted, and that topic now invalidated inside wallet connect bridge, so next try with the same topic, not working?

File: wallet-connect/src/client.rs
Comment (leejw51crypto) [reply wallet-connect/src/client.rs]: yes, it's ok to unwrap or expect.

File: wallet-connect/src/client.rs
Comment (leejw51crypto) [reply wallet-connect/src/client.rs]: i'll change the name, it's for channel message packet.

**Comment by leejw51crypto** [comment]
working on the feedback

File: wallet-connect/src/client/core.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/core.rs]: but in that case, when i rerun web3 example, pending metamask worked

File: wallet-connect/src/client.rs
Comment (tomtau) [reply wallet-connect/src/client.rs]: but why? can a comment be added why it is safe?

File: wallet-connect/src/client/core.rs
Comment (tomtau) [reply wallet-connect/src/client/core.rs]: > user approval not delivered.

I think that's the correct behaviour -- for the persisted session, if it was approved before, it shouldn't need another approval... so persistence, it's ok to save it and assume the state is as before (unless some other update is sent later).
i.e. no need to trigger another approval on each restore... restore should just load the previous session state cached on disk/internal storage

**Comment by tomtau** [comment]
> tried to use same topic, but user approval never delivered

that's expected and that's the correct behaviour: https://github.com/crypto-com/game-sdk-extras/pull/34#discussion_r840262234

File: wallet-connect/src/client/core.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/core.rs]: yes, but it's not approved state, metamask want to sign, but approve page doesn't appear. pending.

i stopped program in qrcode waiting, and ran again.

File: wallet-connect/src/client/session.rs
Comment (tomtau) [reply wallet-connect/src/client/session.rs]: https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.UnboundedSender.html#method.send

> If the receive half of the channel is [close](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.UnboundedReceiver.html#method.close)d, either due to close being called or the [UnboundedReceiver](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.UnboundedReceiver.html) having been dropped, this function returns an error. The error includes the value passed to send.

it can fail -- if the developer who added the callback is not careful and closes or drops the receiver...

Comment (tomtau) [reply wallet-connect/src/client/session.rs]: but I guess if this is hidden from C++ bindings, i.e. C++ bindings always initialise and maintain this stuff correctly, it should be ok

Comment (leejw51crypto) [reply wallet-connect/src/client/session.rs]: ok, i'll add code in case send fails. and add  comment

File: wallet-connect/src/client.rs
Comment (leejw51crypto) [reply wallet-connect/src/client.rs]: i mean, when there are enough memory, and receiver is valid, send will succeed,
and when it fails, it would be process panic state.(out-of-memory etc)

https://docs.rs/futures/0.1.31/futures/sync/mpsc/fn.unbounded.html

I'll add comment

File: wallet-connect/src/client/core.rs
Comment (tomtau) [reply wallet-connect/src/client/core.rs]: do you mean if you shutdown the program while the session state was pending a connection? If that's what you mean, it's better not to save/persist the pending state -- only save/persist the established ones

**Comment by tomtau** [comment]
@leejw51crypto should the client state also store the last/next request ID? Currently, it's an atomic int that gets incremented with every request (i.e. not random)... so when the client is restored, it'll start requests with the IDs that were used for requests before. it's likely fine, but not sure if the bridge server or wallets will have any issue if a duplicate ID is used.

one alternative may also be to generate random IDs instead of incremental

**Comment by leejw51crypto** [comment]
i'll try to include request-id to serialization. check whether it works as expected.

File: wallet-connect/src/client/core.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/core.rs]: yes , no saving  pending state.(just connected, not connected)
once connected, next time, directly goes to personal_sign

**Comment by leejw51crypto** [comment]
added user control (get uri, sessioninfo) for c++ bindings (gamefi)
also that info delivered via callback, too.

File: wallet-connect/src/client.rs
Comment (tomtau) [inline wallet-connect/src/client.rs]: this doesn't give the QR code, but just the connection string (that can be displayed as a QR code)... so the name may be a bit confusing

File: wallet-connect/src/uri.rs
Comment (tomtau) [inline wallet-connect/src/uri.rs]: this seems to be a bit redundant ? one can just call `as_url().to_string()`?

File: wallet-connect/src/client/session.rs
Comment (tomtau) [inline wallet-connect/src/client/session.rs]: the point of having `current_request: AtomicU64,` in `Connector` was that one can access the request id without locking the session... putting a duplicate request id in session will remove this point

**Review by tomtau — CHANGES_REQUESTED** [review]
duplicating the request id Information doesn't seem worth it -- I suggest one of the following:
1. just keep the atomic request id counter as now
2. remove the atomic request id counter and just keep the counter inside session (under the lock)
3. not have id counter, and just use random ids

File: wallet-connect/src/client.rs
Comment (leejw51crypto) [reply wallet-connect/src/client.rs]: ok, i'll change the name

File: wallet-connect/src/uri.rs
Comment (leejw51crypto) [reply wallet-connect/src/uri.rs]: ok, i'll add as comment

**Comment by leejw51crypto** [comment]
removed request id from session, restore back 
use random id in starting (full random u64 no response from server, applied % operation with 4 bytes unsigned range)
renaming

File: wallet-connect/src/client/session.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/session.rs]: i see, i'll remove request-id from session, random will be used instead of 1 .
(to prevent duplicate id)

File: wallet-connect/src/client/core.rs
Comment (tomtau) [inline wallet-connect/src/client/core.rs]: I think the bridge server is implemented in JS, so it may be due to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER

so perhaps just generate u32 and cast it to u64?

Comment (tomtau) [reply wallet-connect/src/client/core.rs]: also as the comment says: "// Trust Wallet requires a non-zero request id" -- perhaps add one to make sure it's non-zero?

File: wallet-connect/src/client/core.rs
Comment (tomtau) [inline wallet-connect/src/client/core.rs]: no need to keep track of this if it's randomly generated... the requests can just generate the id by random, not by accessing this value

File: wallet-connect/src/client/core.rs
Comment (tomtau) [inline wallet-connect/src/client/core.rs]: no need to print?

File: wallet-connect/src/client/session.rs
Comment (tomtau) [inline wallet-connect/src/client/session.rs]: those messages can be constructed inside `if let`?

**Review by tomtau — CHANGES_REQUESTED** [review]
look all right, but it seems the implementation is just doing the random initialisation, but the requests are still incremental... I think if random, it can be random everywhere: so `current_request` field could be removed, and things like https://github.com/crypto-com/game-sdk-extras/blob/f3e428435a8b8fb5afb2a7e3bf5be653c114d3f8/wallet-connect/src/client/core.rs#L181 `self.current_request.fetch_add(1, Ordering::SeqCst);` can be replaced with generating a random id

File: wallet-connect/src/client/core.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/core.rs]: ok

File: wallet-connect/src/client/core.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/core.rs]: i'll remove it, it's debugging code

File: wallet-connect/src/client/core.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/core.rs]: ok

File: wallet-connect/src/client/session.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/session.rs]: i'll check

**Comment by leejw51crypto** [comment]
changed to use random request id,
and non zero , more bandwidth

File: wallet-connect/src/client/session.rs
Comment (tomtau) [inline wallet-connect/src/client/session.rs]: Should the value here depend on `self.info.connected`?

**Review by tomtau — COMMENTED** [review]
ok -- just for the updates... sometimes the update isn't about disconnection, but about changing a network or adding new addresses -- in which case, it's still connected

File: wallet-connect/src/client/session.rs
Comment (leejw51crypto) [reply wallet-connect/src/client/session.rs]: yes, its connecting status is dependent on update.approved.
  ```
  self.info.connected = update.approved;
  ```
  
  i'll change to use `update.approved` 
approved -> call `Connected` 
not approved -> call `Disconnected`

Comment (leejw51crypto) [reply wallet-connect/src/client/session.rs]: changed update(), now it dispatch by the connected status,

and fixed the test,(get_tokens contract information changed)

**Pull request merged**
